File: app/schemas/sector_rotation.py
"""Sector rotation Pydantic schemas.

The contract is intentionally **backward-compatible at the response-shape
level** — endpoint URLs, query params and top-level keys stay the same —
but the semantics of ``sector`` (was: ETF 基金类型 like "股票型"; now:
GICS industry sector like "Information Technology") have changed.

See ``app/services/sector_rotation_service.py`` for the rationale.
"""


from typing import Literal

from pydantic import BaseModel, Field


class SectorScope(BaseModel):
    """Universe scope of the analysis."""

    market: Literal["A股"] = Field(..., description="Trading market scope")
    instrument_types: list[Literal["ETF", "STOCK"]] = Field(
        ..., description="Instrument types included"
    )
    classification: Literal["GICS", "SW"] = Field(
        ...,
        description="Industry classification system: GICS (default, global) "
        "or SW (申万2021一级, A-share only)",
    )


class SectorPerformance(BaseModel):
    """Sector performance metrics (per GICS sector)."""

    sector: str = Field(..., description="GICS sector name (level-1)")
    count: int = Field(..., description="Total instruments in the sector")
    stock_count: int = Field(..., description="Individual stock count")
    etf_count: int = Field(..., description="ETF count")
    return_1w: float = Field(..., description="1-week average return (%)")
    return_1m: float = Field(..., description="1-month average return (%)")
    return_3m: float = Field(..., description="3-month average return (%)")
    return_6m: float = Field(..., description="6-month average return (%)")
    return_1y: float = Field(..., description="1-year average return (%)")
    sharpe_1y: float = Field(..., description="1-year Sharpe (average)")
    volatility_20d: float = Field(..., description="20-day volatility (average)")
    rsi14: float = Field(..., description="RSI14 (average)")
    amount_total: float = Field(..., description="Aggregate turnover (元)")
    relative_strength_1w: float = Field(
        ..., description="Relative strength vs market avg (1w)"
    )
    relative_strength_1m: float = Field(
        ..., description="Relative strength vs market avg (1m)"
    )
    relative_strength_3m: float = Field(
        ..., description="Relative strength vs market avg (3m)"
    )
    momentum_rank: int = Field(..., description="Rank by 1m return (1=best)")


class RotationSignal(BaseModel):
    """Sector rotation signal — sector moved up/down ≥3 positions in rank."""

    sector: str
    type: str  # "up" or "down"
    message: str
    current_rank: int
    previous_rank: int
    rank_change: int = Field(
        ..., description="Positive = moved up; negative = moved down"
    )


class MarketAverage(BaseModel):
    """Market-average metrics across all sectors in scope."""

    return_1w: float
    return_1m: float
    return_3m: float
    return_6m: float
    return_1y: float
    sharpe_1y: float


class SectorRotationResponse(BaseModel):
    """Sector rotation analysis response."""

    trade_date: str
    scope: SectorScope
    sectors: list[SectorPerformance]
    market_avg: MarketAverage | None
    rotation_signals: list[RotationSignal]


class SectorListItem(BaseModel):
    """Sector list item — distinct sector with composition counts."""

    sector: str
    count: int
    stock_count: int
    etf_count: int


class SectorListResponse(BaseModel):
    """Sector list response."""

    items: list[SectorListItem]


# ---------------------------------------------------------------------------
# Constituents view (added 2026-07-09)
# ---------------------------------------------------------------------------


class SectorConstituent(BaseModel):
    """Single instrument inside a sector, ranked by weight (market cap for
    STOCK, fund size for ETF).

    The ``weight`` field is intentionally heterogeneous — the
    ``weight_unit`` and ``weight_label`` fields tell the UI how to render
    it. Mixing STOCKs (市值 in 元) and ETFs (规模 in 元) in one table is
    fine for drill-down purposes: the user is looking at the *composition*
    of the sector, not a weighted sum.
    """

    code: str
    name: str
    instrument_type: Literal["ETF", "STOCK"]
    # Sector classification that resolved (might be `etf_info.sector` or
    # the keyword-inferred fallback for ETFs). Echoed so the client can
    # render a tag without re-running the heuristic.
    resolved_sector: str
    # Weight metric. None when the upstream data is missing (e.g. stock
    # not yet in `stock_fundamental`).
    weight: float | None = Field(
        None, description="Weight value (market cap or fund size in CNY 元)"
    )
    weight_unit: Literal["元"] = Field(
        "元", description="Currency unit for weight (always CNY 元)"
    )
    weight_label: Literal["市值", "规模"] = Field(
        ...,
        description="Label for the weight column: '市值' for STOCK, '规模' for ETF",
    )

    # Period returns — same shape as SectorPerformance so the UI can
    # reuse the ReturnTag component.
    return_1w: float | None = None
    return_1m: float | None = None
    return_3m: float | None = None
    return_6m: float | None = None
    return_1y: float | None = None

    # Liquidity / quality (optional — surfaced when present)
    sharpe_1y: float | None = None
    rsi14: float | None = None
    amount_total: float | None = None


class SectorConstituentsResponse(BaseModel):
    """Top-N constituents for a single GICS sector.

    Returned by ``GET /sector-rotation/sectors/{sector}/constituents``.
    """

    sector: str = Field(..., description="GICS sector name (matches path param)")
    trade_date: str | None = Field(
        None, description="Analysis date (ISO); null when no indicators are seeded"
    )
    count: int = Field(..., description="Number of constituents returned")
    total_in_sector: int = Field(
        ...,
        description=(
            "Total instruments in the sector (after sector-resolution). "
            "Useful for the UI to render 'showing top N of M'."
        ),
    )
    items: list[SectorConstituent]

File: web/src/types/sector_rotation.ts
/**
 * Sector rotation domain types.
 *
 * Re-exports the API client types so feature code can import from
 * `@/types/sector_rotation` (the project's convention — see
 * `@/types/instrument`, `@/types/screen`).
 *
 * Field-level docs live in `app/schemas/sector_rotation.py`.
 */

/** Industry classification system: GICS (global default) or 申万 (A-share). */
export type SectorClassification = 'GICS' | 'SW';

export interface SectorScope {
  market: 'A股';
  instrument_types: Array<'ETF' | 'STOCK'>;
  classification: SectorClassification;
}

export interface SectorPerformance {
  sector: string;
  count: number;
  stock_count: number;
  etf_count: number;
  return_1w: number;
  return_1m: number;
  return_3m: number;
  return_6m: number;
  return_1y: number;
  sharpe_1y: number;
  volatility_20d: number;
  rsi14: number;
  amount_total: number;
  relative_strength_1w: number;
  relative_strength_1m: number;
  relative_strength_3m: number;
  momentum_rank: number;
  // ----------------------------------------------------------------
  // Phase 3 (added 2026-07-19): SW 模式下，return_* 已切换为申万一级
  // 指数官方回报；constituent_return_* 保留等权 ETF+STOCK 回报作为
  // 对照。return_source 标记当前 return_* 的来源。
  //
  // - return_source: "official_index"           — return_* 来自官方指数
  //                         (SW 模式 + sw_industry_index_return 有行)
  //                 : "constituents_equal_weight" — return_* 来自等权平均
  //                         (GICS 模式或 SW 模式但缺官方行)
  // - sw_l1_code: 申万一级行业代码 (e.g. "801080"), 仅 SW 模式有
  // - official_close: 申万指数当日收盘点位, 仅 SW 模式有
  // - constituent_return_*: 等权回报对照, 仅 SW 模式有
  // ----------------------------------------------------------------
  return_source?: 'official_index' | 'constituents_equal_weight';
  sw_l1_code?: string | null;
  official_close?: number | null;
  constituent_return_1w?: number | null;
  constituent_return_1m?: number | null;
  constituent_return_3m?: number | null;
  constituent_return_6m?: number | null;
  constituent_return_1y?: number | null;
}

export interface RotationSignal {
  sector: string;
  type: 'up' | 'down';
  message: string;
  current_rank: number;
  previous_rank: number;
  rank_change: number;
}

export interface MarketAverage {
  return_1w: number;
  return_1m: number;
  return_3m: number;
  return_6m: number;
  return_1y: number;
  sharpe_1y: number;
}

export interface SectorRotationData {
  trade_date: string | null;
  scope: SectorScope;
  sectors: SectorPerformance[];
  market_avg: MarketAverage | null;
  rotation_signals: RotationSignal[];
}

export interface SectorListItem {
  sector: string;
  count: number;
  stock_count: number;
  etf_count: number;
}

export interface SectorListData {
  items: SectorListItem[];
}

// ---------------------------------------------------------------------------
// Constituents view (added 2026-07-09)
// ---------------------------------------------------------------------------

export type SectorConstituentType = 'ETF' | 'STOCK';

export interface SectorConstituent {
  code: string;
  name: string;
  instrument_type: SectorConstituentType;
  /** Resolved GICS sector (echoed from the server). */
  resolved_sector: string;
  /** Weight in CNY 元. null when the upstream data is missing. */
  weight: number | null;
  weight_unit: '元';
  /** '市值' for STOCK, '规模' for ETF — drives the column header. */
  weight_label: '市值' | '规模';
  return_1w: number | null;
  return_1m: number | null;
  return_3m: number | null;
  return_6m: number | null;
  return_1y: number | null;
  sharpe_1y: number | null;
  rsi14: number | null;
  amount_total: number | null;
}

export interface SectorConstituentsData {
  sector: string;
  trade_date: string | null;
  count: number;
  total_in_sector: number;
  items: SectorConstituent[];
}

/** Display columns used by the heatmap & detail table. */
export type SectorReturnPeriod = '1w' | '1m' | '3m' | '6m' | '1y';

export const SECTOR_RETURN_PERIODS: SectorReturnPeriod[] = ['1w', '1m', '3m', '6m', '1y'];

export const SECTOR_RETURN_LABELS: Record<SectorReturnPeriod, string> = {
  '1w': '1周',
  '1m': '1月',
  '3m': '3月',
  '6m': '6月',
  '1y': '1年',
};